### ISSUE TITLE
helm-get-org-candidates-in-file truncates headings at the wrong position

### ISSUE THREAD

**Issue by tmalsburg** [emacs-helm/helm#1294]
`window-width` is used to determine the point for truncation but the function doesn't use the window width of the helm-window but that of the window in which helm was called.

**Comment by thierryvolpiatto** [comment]
Thanks to look into this, I have not implemented this myself and BTW I
am not familiar with helm-org.el, anyway I will have a look ASAP,
possibly on thursday, otherwise at end of the week.

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by tmalsburg** [comment]
See also my comments on #964.  Thank you Thierry!

**Comment by thierryvolpiatto** [comment]
Titus von der Malsburg <notifications@github.com> writes:

> See also my comments on #964. Thank you Thierry!

I tried to improve all this, hope that fixed your problems.

Thanks.

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by jagrg** [comment]
With this recent update (`helm-20151218.329`), I'm no longer able to change min and max-depth on `helm-source-org-headings-for-files`. I used to be able to set for example (min-depth 2) (max-depth 2) and it would only display level 2 headlines.

**Comment by thierryvolpiatto** [comment]
Jonathan Gregory <notifications@github.com> writes:

> With this recent update (helm-20151218.329), I'm no longer able to
> change min and max-depth on helm-source-org-headings-for-files. I used
> to be able to set for example (min-depth 2) (max-depth 2) and it would
> only display level 2 headlines.

Please clarify how you are using this, with a recipe if possible.

Thanks.

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by jagrg** [comment]
I was using it to search bibliographical notes. See: https://gist.github.com/jagrg/39d68432c783df50d199.

**Comment by thierryvolpiatto** [comment]
Jonathan Gregory <notifications@github.com> writes:

> I was using it to search bibliographical notes. See: https://gist.github.com/jagrg/39d68432c783df50d199.

Why are you using:

(cl-defun helm-bibtex-source-org-headings-for-files (filenames &optional
                                                    (min-depth 2) (max-depth 2))

instead of simply passing 2 (min-depth) and 2 (max-depth) to
helm-org-get-candidates ?

I can't try your code because I have no idea how to install helm-bibtex
and its dependencies and no sample file, however the min-depth and
max-depth arguments (1 and 8 by default) are passing correctly here (I
could modify them without problems)

Also be sure to recompile correctly all your files.
 
-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by tmalsburg** [comment]
Matching is apparently still done on the truncated headlines.  I used the following org mode file for testing:

```
* Arizona
* sdlkgj dslkjg sldkjg lskdjg lskdjg lksdjg lksdjg lksdjg lskdj glskdjg lskdjg lskdjg lskdjg lskdjg lskdjg lksdjg lskdjg lksdjg lksdjg lksdj glksjd glksjd glskdj glskdjg lskdjglksjdg lks Arizona
```

When I fire up `helm-org-in-buffer-headings` and search for "Arizona", I only get a match for the first headline.  Tested with 76111339c00731c97e917f35a3121f3f294472d8.

**Comment by tmalsburg** [comment]
@jagrg your issue is probably independent from helm-bibtex.  You could make Thierry's life easier by constructing a minimal reproducible example.

**Comment by thierryvolpiatto** [comment]
Titus von der Malsburg <notifications@github.com> writes:

> When I fire up helm-org-in-buffer-headings and search for "Arizona", I
> only get a match for the first headline. Tested with 7611133.

Good, thanks for the recipe, will look also into this soon.


-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by jagrg** [comment]
@thierryvolpiatto, here's a test file: https://gist.github.com/jagrg/6305b022557b80d7f012. You don't need helm-bibtex installed. This is what you need to do:

1) `(setq my-test-file "~/path/to/test.org")`
2) You can use any org file with level 2 headlines, but suppose you have something like:
```
#+TITLE: Helm test

* Methods

** Skinner, J. (2012) The interview: an ethnographic approach
Hello world.

** Burgess, R. (1991) In the field: an introduction to field research
The quick brown fox jumps over the lazy dog.
```
3) Now do: `M-x helm-test-show-notes`

This is what I see:
```
Methods/Skinner, J. (2012) The interview: an ethnographic approach
Methods/Burgess, R. (1991) In the field: an introduction to field research
```
And this is what I used to see before the commits:
```
Skinner, J. (2012) The interview: an ethnographic approach
Burgess, R. (1991) In the field: an introduction to field research
```

**Comment by thierryvolpiatto** [comment]
Thanks for the recipe.

Jonathan Gregory <notifications@github.com> writes:

> This is what I see:
>
> Methods/Skinner, J. (2012) The interview: an ethnographic approach
> Methods/Burgess, R. (1991) In the field: an introduction to field research

Yes this is what is expected.

> And this is what I used to see before the commits:
>
> Skinner, J. (2012) The interview: an ethnographic approach
> Burgess, R. (1991) In the field: an introduction to field research

This is not related to last commits AFAICT, this was the behavior some months
ago, perhaps you haven't updated helm since a long time ?

Did you try to pass your arguments directly to helm-org-get-candidates ?

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by jagrg** [comment]
This is the commit where the problem begins: https://github.com/emacs-helm/helm/commit/d6c8b7c86c3c32b03adec2e2e223c9a4c5da0b3e

I know this is the desired behaviour but I'm wanting to control exactly which level to display. In my case, that's level 2. I was hoping that changing min and max-depth would work (and it did before the commit above). I also tried to change `helm-org-get-candidates` as you suggested but it didn't work. Is this what you were thinking:

```emacs-lisp
(cl-defun helm-org-get-candidates (filenames min-depth max-depth &optional (parents nil))
  (helm-flatten-list
   (mapcar (lambda (filename)
             (helm-org--get-candidates-in-file
              filename 2 2
              helm-org-headings-fontify
              (if parents t helm-org-headings--nofilename)
              parents))
           filenames)))
```

**Comment by thierryvolpiatto** [comment]
Jonathan Gregory <notifications@github.com> writes:

> This is the commit where the problem begins: d6c8b7c
>
> I know this is the desired behaviour but I'm wanting to control
> exactly which level to display. In my case, that's level 2.

So that's not what your recipe showed, anyway I am here able to change
in source code the value of min-depth and max-depth.

BTW your recipe is not usable, please give me a recipe I can use.

> I was hoping that changing min and max-depth would work (and it did
> before the commit above).

    (cl-defun helm-source-org-headings-for-files
        (filenames
         &optional (min-depth 1) (max-depth 8) parents)

If I change here min-depth and max-depth it works as expected, arguments
are passed.

Be sure to use lexical-binding though.

> I also tried to change helm-org-get-candidates as you suggested but it
> didn't work. Is this what you were thinking:

No, I meant in your code.

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by thierryvolpiatto** [comment]
Here example with screenshots:

With actual code, I run `helm-org-in-buffer-headings` from the test.org buffer:

![screenshot - 19122015 - 06 35 13](https://cloud.githubusercontent.com/assets/1533939/11911609/4f13f9e6-a61b-11e5-8224-79f7979d202f.png)

As you can see all headings are there.
Now I eval in source code `helm-source-org-headings-for-files` with its default arguments modified:

![screenshot - 19122015 - 06 36 12](https://cloud.githubusercontent.com/assets/1533939/11911617/9155e8f0-a61b-11e5-87f8-5191543a7766.png)

And now I come back to test.org and I run again `helm-org-in-buffer-headings`:

![screenshot - 19122015 - 06 36 50](https://cloud.githubusercontent.com/assets/1533939/11911622/c54592c8-a61b-11e5-92f7-b1f90eb0328c.png)

As you can see only heading with 2 level is shown.

**Comment by jagrg** [comment]
Thanks for the explanation, however you are missing my point (and BTW sorry for not being clear enough). This is what your last screenshot shows:

`test1/test2`

And this is what I am trying to accomplish (after changing the depth, of course):

`test2`

If you change min and max-depth prior to https://github.com/emacs-helm/helm/commit/d6c8b7c86c3c32b03adec2e2e223c9a4c5da0b3e you should see that only level 2 is shown (without the parent headline).

In any case, considering the size of the monitor, this becomes especially problematic when using long headlines. Lower level headlines are sometimes unreadable. One solution (just food for through) would be to use an outline structure instead, similar to org, which IMO is a lot cleaner. So that this:

```
My first parent headline
My first parent headline/My second level headline
My first parent headline/My second level headline/My third level headline
My first parent headline/My second level headline/My third level headline/My fourth level headline
My second parent headline
My second parent headline/My second level headline
My second parent headline/My second level headline/My third level headline
My second parent headline/My second level headline/My third level headline/My fourth level headline
```

becomes this:

```
My first parent headline
 My second level headline
  My third level headline
   My fourth level headline
My second parent headline
 My second level headline
  My third level headline
   My fourth level headline
```

**Comment by thierryvolpiatto** [comment]
Jonathan Gregory <notifications@github.com> writes:

> Thanks for the explanation, however you are missing my point (and BTW
> sorry for not being clear enough).

Ok I understand now,
I will look on monday what makes the difference, hope this will be fixed
before the end of the year ;-)

> This is what your last screenshot shows:
>
> test1/test2
>
> And this is what I am trying to accomplish (after changing the depth, of course):
>
> test2
>
> If you change min and max-depth prior to d6c8b7c you should see that
> only level 2 is shown (without the parent headline).

Ok that's clear.

> In any case, considering the size of the monitor, this becomes
> especially problematic when using long headlines. Lower level
> headlines are sometimes unreadable. One solution (just food for
> through) would be to use an outline structure instead, similar to org,
> which IMO is a lot cleaner. So that this:

PR's welcome, but IMO the best structure to read org files is org
itself.
Doing this is probably cleaner for parsing one file, but with many org
files (with a lot of headers), people generally want to see the path and
the differents levels.

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by thierryvolpiatto** [comment]
Jonathan Gregory <notifications@github.com> writes:

> If you change min and max-depth prior to d6c8b7c you should see that
> only level 2 is shown (without the parent headline).

I observe the same behavior after reverting all the changes I did
recently, i.e I still have "test1/test2".

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by jagrg** [comment]
That's strange. I was even able to reproduce it in `emacs -Q`. Did you restart emacs?

Try this:

1. make sure you have the latest melpa version installed, which is `helm-20151218.1216`
2. now open the `helm-org` library, delete its content and paste [this code](https://raw.githubusercontent.com/emacs-helm/helm/62f93148fa2ab95d32d4bf50753cc66184760a81/helm-org.el), which is the version previous to when the problem was introduced.
3. change min and max-depth to 2 (on line 60 of helm-org.el) and save the file.
4. recompile `helm-20151218.1216` by typing `M-x byte-recompile-directory`.
5. restart emacs
6. open the test.org file and type `M-x helm-org-in-buffer-headings`

What do you see?

**Comment by thierryvolpiatto** [comment]
Jonathan Gregory <notifications@github.com> writes:

> What do you see?

I finally was able to reproduce by stepping in all recents changes, it
appear I was doing exactly the contrary of what was done initially to
call org-get-outline-path.
Should be fixed now.
Thanks.

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by jagrg** [comment]
Thanks. It does work, but there's an issue with changing the depth back and forth, which I think could become a useful feature for cycling through headline levels. Anyhow, here's the problem. Add the following to your test.org file:

```
* test1
** test2
*** test3
**** test4
* testA
** testB
*** testC
**** testD
```

1. restart emacs
2. open test.org and type `M-x helm-org-in-buffer-headings`
3. note that the full path is displayed as expected
4. now open helm-org.el and change min and max-depth to 2
5. save file, `eval-buffer` and recompile
6. repeat step 2

This is what I see:

```
testA/test2
testA/testB
```

Now, apart from the path which I think shouldn't be there, `testA` is not a parent of `test2`.

**Comment by thierryvolpiatto** [comment]
Jonathan Gregory <notifications@github.com> writes:

> Now, apart from the path which I think shouldn't be there, testA is not a parent of test2.

Indeed, is that new of this bug was reproductible in previous versions ?

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by jagrg** [comment]
I wasn't able to reproduce with https://github.com/emacs-helm/helm/commit/e42e3ed6543b72756e6f2f1a96cbd0d35166cb6e, which I believe is the version prior to when the depth issue was addressed. As you can see, the paths are correct:

```
test1/test2
testA/testB
```

**Comment by jagrg** [comment]
Changing `unless` to `when` on [line 143](https://github.com/emacs-helm/helm/blob/master/helm-org.el#L143) fixes the parent heading problem. I'm not sure if this would have any implication on the rest of the code though.

**Comment by thierryvolpiatto** [comment]
No, this work for you with 2 and 2, but this is not what the original code
wants to do.
BTW the actual code works perfectly with 2 and 2 if you recompile and
restart emacs.
Le 22 déc. 2015 14:00, "Jonathan Gregory" <notifications@github.com> a
écrit :

**Comment by jagrg** [comment]
As I said, yes, your code does work when changing the source code directly and you are welcome to close this. Helm is tricky though. I haven't quite figured out how to create multiple `helm-org-in-buffer-headings` functions with different min and max-depth arguments, like:

`M-x helm-org-in-buffer-headings` --> display the entire headline structure and path
`M-x helm-org-in-buffer-headings-1` --> display level 1 only (without the path)
`M-x helm-org-in-buffer-headings-2` --> display level 2 only (without the path)
etc.

and eventually a function to cycle through the headline levels back and forth, which I think would make 
`helm-org` a lot more potent, especially for long org files with many layers. Indeed, I never really liked the way `helm-org-in-buffer-headings` displays the full path, which sometimes makes filtering less effective due to the repetition of words. In any case, I probably wouldn't have the knowledge for a PR, but the outline structure which I mentioned earlier is likely the best option. Maybe something to consider at some point.

**Comment by thierryvolpiatto** [comment]
Closing now as all issues are now fixed, please see #1300 for further changes.

**Comment by thierryvolpiatto** [comment]
Jonathan Gregory <notifications@github.com> writes:

> As I said, yes, your code does work when changing the source code
> directly and you are welcome to close this. Helm is tricky though. I
> haven't quite figured out how to create multiple
> helm-org-in-buffer-headings functions with different min and max-depth
> arguments, like:
>
> M-x helm-org-in-buffer-headings --> display the entire headline structure and path
> M-x helm-org-in-buffer-headings-1 --> display level 1 only (without the path)
> M-x helm-org-in-buffer-headings-2 --> display level 2 only (without the path)
> etc.

You can now use helm-org-headings-min/max-depth, HTH.

> and eventually a function to cycle through the headline levels back
> and forth, which I think would make helm-org a lot more potent,
> especially for long org files with many layers. Indeed, I never really
> liked the way helm-org-in-buffer-headings displays the full path,

Me too, it is now configurable and the default is no path.

> 
> which sometimes makes filtering less effective due to the repetition
> of words.

Indeed.

Thanks.

-- 
Thierry
https://emacs-helm.github.io/helm/
